### PR TITLE
Support for YOLOv7 ONNX (not simplified)

### LINKED ISSUE #22215: When support to YOLO v7?
- OpenCV => 4.6.0
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2017
-->

- Compiler => :grey_question:

##### Detailed description

How support to YOLOv7 by onnx format?

https://github.com/WongKinYiu/yolov7

### PR DESCRIPTION
**Pull request by fengyuentau**
Fixes #22215 .

Test data: https://github.com/opencv/opencv_extra/pull/998
Link to the model: https://drive.google.com/u/0/uc?id=1rm3mIqjJNu0xPTCjMKnXccspazV1B2zv&export=download

Things have to be done for the support:
- [x] Identity node: getting input from constBlob.
- [x] [Range node](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Range): implementation for Range node (all inputs are constant).
- [x] fix parseExpand with multiple broadcast axes.
- [x] fix 1D mat issue comes with Range output.
- [x] acc test

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
buildworker:Custom=linux-f1
build_image:Custom=ubuntu:18.04
```

### CODE REVIEW

**Comment by KishoreElvicto** [comment]
kindly fix this issue as soon as possible

**Comment by fengyuentau** [comment]
All is ready for reviews. Let me test other linked issues as well.

File: modules/dnn/test/test_model.cpp [hunk 1]
```diff
@@ -810,8 +810,8 @@ TEST_P(Test_Model, TextDetectionByEAST)
 
 TEST_P(Test_Model, YOLOv7)
 {
-    std::string weightPath = findDataFile("yolov7-not-simplified.onnx", true);
-    std::string imgPath = findDataFile("test5.jpg", true);
+    std::string weightPath = _tf("onnx/models/yolov7_not_simplified.onnx");
```
Comment (peters) [inline modules/dnn/test/test_model.cpp:813]: @fengyuentau Which export arguments from the `yolov7` repository did you use to produce a working onnx model? :)

Comment (fengyuentau) [reply modules/dnn/test/test_model.cpp:813]: This model is exported using the script from https://github.com/opencv/opencv/issues/22215#issuecomment-1190053395, which is not simplified so that some changes need to be done for OpenCV DNN to support.

You can get a carefully exported YOLOv7 already working with OpenCV DNN from [this comment](https://github.com/opencv/opencv/issues/22215#issuecomment-1190554236).

Comment (peters) [reply modules/dnn/test/test_model.cpp:813]: Thank you! :)

**Review by zihaomu — COMMENTED** [review]
Thanks for your contribution!
Please check the comments below, others are fine for me.

File: modules/dnn/test/test_model.cpp [hunk 2]
```diff
@@ -5,6 +5,8 @@
 #include "test_precomp.hpp"
 #include <opencv2/dnn/shape_utils.hpp>
 #include "npy_blob.hpp"
+#include <opencv2/core/utils/logger.hpp>
```
Comment (zihaomu) [inline modules/dnn/test/test_model.cpp:8]: Import unused header files

Comment (fengyuentau) [reply modules/dnn/test/test_model.cpp:8]: fixed.

File: modules/dnn/test/test_model.cpp [hunk 3]
```diff
@@ -806,6 +807,88 @@ TEST_P(Test_Model, TextDetectionByEAST)
...
+    for (auto i : keep_idx)
+    {
+        keep_classIds.push_back(classIds[i]);
+        keep_confidences.push_back(confidences[i]);
+        keep_boxes.push_back(boxes[i]);
+    }
+
+    //CV_LOG_INFO(NULL, "det0: " << "id=" << keep_classIds[0] << ", conf=" << keep_confidences[0] << ", box=" << keep_boxes[0]);
+    //CV_LOG_INFO(NULL, "det1: " << "id=" << keep_classIds[1] << ", conf=" << keep_confidences[1] << ", box=" << keep_boxes[1]);
+    //CV_LOG_INFO(NULL, "det2: " << "id=" << keep_classIds[2] << ", conf=" << keep_confidences[2] << ", box=" << keep_boxes[2]);
```
Comment (zihaomu) [inline modules/dnn/test/test_model.cpp:888]: Maybe we need to remove this commented code or keep the commented version of `std::cout`.

Comment (fengyuentau) [reply modules/dnn/test/test_model.cpp]: Done removing.

**Review by zihaomu — APPROVED** [review]
LGTM! 👍

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 4]
```diff
@@ -3091,6 +3108,61 @@ void ONNXImporter::parseDepthToSpace(LayerParams& layerParams, const opencv_onnx
...
+    {
+        std::vector<Mat> input, output;
+        input.push_back(getBlob(node_proto, 0));
+        runLayer(layerParams, input, output);
+        addConstant(node_proto.output(0), output[0]);
+        return;
+    }
+
+    addLayer(layerParams, node_proto);
+}
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_importer.cpp:3128]: I think we could move this code to parseSimpleLayers with minimal changed and we get support for more constant layer, what do you think?

**Comment by fengyuentau** [comment]
For https://github.com/opencv/opencv/issues/22361, the user does not provide the model.

For https://github.com/opencv/opencv/issues/20324, this patch solved the Range support. But DNN still have problems dealing with other operators in that model.

File: modules/dnn/test/test_model.cpp [hunk 5]
```diff
@@ -806,6 +807,85 @@ TEST_P(Test_Model, TextDetectionByEAST)
...
+
+    // Reference, which is collected with input size of 640x640
+    std::vector<int> refClassIds{1, 16, 7};
+    std::vector<float> refScores{0.9624768f, 0.95940655f, 0.8693003f};
+    // [x1, y1, x2, y2] x 3
+    std::vector<Rect2d> refBoxes{Rect2d(105.99881f, 150.2002f,  472.52933f, 466.508f),
+                                  Rect2d(109.99532f, 246.13585f, 259.82266f, 600.7827f),
+                                  Rect2d(386.13947f, 83.04132f,  575.3909f,  189.83206f)};
+
+    Mat img = imread(imgPath);
```
Comment (rogday) [inline modules/dnn/test/test_model.cpp:827]: Can `testDetectModel` be used here?

**Review by rogday — COMMENTED** [review]
Thank you for contribution! Looks good, have just a few questions.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 4 above)
Comment (fengyuentau) [reply modules/dnn/src/onnx/onnx_importer.cpp:3128]: I would say it should be a rare case for other layers to take a constant as input. But in terms of better compatibility, let me do this.

Layers that trigger parseSimpleLayers:
https://github.com/opencv/opencv/blob/42561e1233659c30ab8f88b49212122984f6d218/modules/dnn/src/onnx/onnx_importer.cpp#L3632-L3635

File: modules/dnn/test/test_model.cpp (hunk 5 above)
Comment (fengyuentau) [reply modules/dnn/test/test_model.cpp:827]: Let me have a try.

Comment (fengyuentau) [reply modules/dnn/test/test_model.cpp:827]: I just recalled a reason why I did not use `testDetectModel`. For YOLOv7, the final confidence for a box is calculated by multiplying the class confidence and the object confidence in line 855 (`conf *= obj_conf;`). I dont see this in `testDetectModel.detect()`, so I did not use `testDetectModel`.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 4 above)
Comment (fengyuentau) [reply modules/dnn/src/onnx/onnx_importer.cpp]: Done.

**Comment by fengyuentau** [comment]
I will try manually upload the model to each CI server tomorrow.

File: modules/dnn/test/test_model.cpp [hunk 6]
```diff
@@ -806,6 +807,85 @@ TEST_P(Test_Model, TextDetectionByEAST)
...
+            continue;
+
+        // get class id and conf
+        Mat scores = preds.row(i).colRange(5, preds.cols);
+        double conf;
+        Point maxLoc;
+        minMaxLoc(scores, 0, &conf, 0, &maxLoc);
+        conf *= obj_conf;
+        if (conf < conf_threshold)
+            continue;
```
[the remark this answers is not in the document]
Comment (fengyuentau) [reply modules/dnn/test/test_model.cpp:857]: To my understanding, recent object detection models (e.g. [YOLOX in opencv_zoo](https://github.com/opencv/opencv_zoo/pull/86/files#diff-e00ac7b5ef6f02331aac9d5b854ce37b7f6c8cbaa60aeb376a8c5bec75b26f07R70)) are trained with predicting IoU between anchors and ground truth boxes ([related lines in YOLOv7 repo](https://github.com/WongKinYiu/yolov7/blob/44d8ab41780e24eba563b6794371f29db0902271/utils/loss.py#L468-L472)). In this case, the final confidence should be calculated by multiplying the class confidence and the predicted IoU (which is named `obj_conf` in my code).

In short, models without such a training scheme should not do this multiplication. So I do not think it makes sense if we do this for region/detection_output layer, considering compatibility for old object detection models.

By the way, we should consider support next gen NMS algorithms (CIoU, DIoU ...) since they are quite popular among recent object detection models.

**Comment by fengyuentau** [comment]
> I will try manually upload the model to each CI server tomorrow.

Done. GitHub Actions tests are all passed.

**Comment by zihaomu** [comment]
Hi @fengyuentau, please take a look [this comment](https://github.com/opencv/opencv/issues/22446#issuecomment-1233649431). This patch can work correct on `Simple Layer`, but it maybe cause some issue in `reshape layer`. [Test model link](https://github.com/opencv/opencv/files/9466261/model_2022_0714_10h14_12.onnx.zip)

**Comment by fengyuentau** [comment]
Hello @alalek , could kindly you update the dnn model in default? Link to the model is attached to the first comment.

**Comment by alalek** [comment]
There is some problem with script:

```
Model <yolov7_not_simplified>
  expect fcd0fa401c83bf2b29e18239a9c2c989c9b8669d
  catch [Errno 2] No such file or directory: 'onnx/models/yolov7_not_simplified.onnx'
  hash check failed - downloading
  get https://drive.google.com/u/0/uc?id=1rm3mIqjJNu0xPTCjMKnXccspazV1B2zv&export=download
  200 OK [<unknown> Mb]
  progress > done
 file onnx/models/yolov7_not_simplified.onnx
  expect fcd0fa401c83bf2b29e18239a9c2c989c9b8669d
  actual bc7f7fe19f90edbde1f71f638c49a16bc2ebe7db
  renaming invalid file to onnx/models/yolov7_not_simplified.onnx.invalid
Following models have not been downloaded:
* onnx/models/yolov7_not_simplified.onnx
```

For Google documents we should use `GDrive()` (see download_models.py)

**Comment by alalek** [comment]
There are 10 commits, some of them are not needed at all in git history

> delete useless files

So please squash commits into one (1 PR - 1 commit rule).

**Comment by fengyuentau** [comment]
@alalek Thank you for the feedback and suggestions. Commits are squashed into one and the model is now downloaded with GDrive() in opencv_extra.

**Comment by fengyuentau** [comment]
> Update2: File is downloaded manually.
GDrive() downloader would be improved later (starting from 3.4 branch).

Thank you for the information! I thought I use GDrive() in a wrong way and was testing it till now.

**Comment by alalek** [comment]
There are failures of test accuracy checks on [CI](https://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/101043) (i5-6600 through KVM):

```
[ RUN      ] Test_Model.YOLOv7/0, where GetParam() = OCV/CPU
[ INFO:0@228.119] global /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (822) populateNet DNN/ONNX: loading ONNX v7 model produced by 'pytorch':1.12.0. Number of nodes = 576, initializers = 193, inputs = 1, outputs = 4
[ INFO:0@228.119] global /build/precommit_linux64/4.x/opencv/modules/dnn/src/onnx/onnx_importer.cpp (738) parseOperatorSet DNN/ONNX: ONNX opset version = 13
Unmatched prediction: class 1 score 0.962365 box [472.545 x 466.494 from (106.011, 150.323)]
Highest IoU: 0
/build/precommit_linux64/4.x/opencv/modules/dnn/test/test_common.impl.hpp:143: Failure
Value of: matched
  Actual: false
Expected: true
Unmatched prediction: class 16 score 0.959694 box [259.837 x 600.703 from (110.007, 246.323)]
Highest IoU: 0
/build/precommit_linux64/4.x/opencv/modules/dnn/test/test_common.impl.hpp:143: Failure
Value of: matched
  Actual: false
Expected: true
Unmatched prediction: class 7 score 0.879178 box [575.834 x 189.781 from (386.032, 83.0033)]
Highest IoU: 0
/build/precommit_linux64/4.x/opencv/modules/dnn/test/test_common.impl.hpp:143: Failure
Value of: matched
  Actual: false
Expected: true
Unmatched reference: class 1 score 0.962477 box [472.529 x 466.508 from (105.999, 150.2)] IoU diff: 1
/build/precommit_linux64/4.x/opencv/modules/dnn/test/test_common.impl.hpp:155: Failure
Expected: (refScores[i]) <= (confThreshold), actual: 0.962477 vs 0
Unmatched reference: class 16 score 0.959407 box [259.823 x 600.783 from (109.995, 246.136)] IoU diff: 1
/build/precommit_linux64/4.x/opencv/modules/dnn/test/test_common.impl.hpp:155: Failure
Expected: (refScores[i]) <= (confThreshold), actual: 0.959407 vs 0
Unmatched reference: class 7 score 0.869300 box [575.391 x 189.832 from (386.139, 83.0413)] IoU diff: 1
/build/precommit_linux64/4.x/opencv/modules/dnn/test/test_common.impl.hpp:155: Failure
Expected: (refScores[i]) <= (confThreshold), actual: 0.8693 vs 0
[  FAILED  ] Test_Model.YOLOv7/0, where GetParam() = OCV/CPU (3427 ms)
```

**Comment by fengyuentau** [comment]
> There are failures of test accuracy checks on [CI](https://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/101043) (i5-6600 through KVM):

@alalek I cannot reproduce this issue on my side, and we do not have these failures in GHA checks. I looked into the error message, and the problem was the difference between refScores and testScores (0.000111818, 0.00028795, 0.00987768) exceeding the threshold (1e-5). So these differences break the condition in line 124, and all IoU calculations are skipped, leading to all boxes mismatched:
https://github.com/opencv/opencv/blob/3456d28cc2d52989f6a8d05ffb35765b6cbbb028/modules/dnn/test/test_common.impl.hpp#L121-L136

I even tried manually update the base of my branch, re-compile everything, but still the check was passed and the issue cannot be reproduced.

File: modules/dnn/test/test_model.cpp [hunk 7]
```diff
@@ -806,6 +807,85 @@ TEST_P(Test_Model, TextDetectionByEAST)
     );
 }
 
+TEST_P(Test_Model, YOLOv7)
+{
+    std::string weightPath = _tf("onnx/models/yolov7_not_simplified.onnx");
+    std::string imgPath = _tf("dog_orig_size.jpg");
```
Comment (alalek) [inline modules/dnn/test/test_model.cpp:813]: JPEG or other loss image formats should not be used in tests as input data.

I have at least 3 configurations with different CRC64 values. Which one has gold values?

File: modules/dnn/test/test_model.cpp [hunk 8]
```diff
@@ -5,6 +5,7 @@
 #include "test_precomp.hpp"
 #include <opencv2/dnn/shape_utils.hpp>
 #include "npy_blob.hpp"
+
```
Comment (alalek) [inline modules/dnn/test/test_model.cpp:8]: > modules/dnn/test/test_model.cpp

Wrong test file is chosen.
This file is used for testing of "DNN Model API".

Darknet models should go into `test_darknet_importer.cpp` file.
If it is not related to Darknet framework, then it should got into `test_onnx_importer.cpp`.

This is how it is organized - anyway we don't need extra mess.

File: modules/dnn/test/test_model.cpp [hunk 9]
```diff
@@ -806,6 +807,85 @@ TEST_P(Test_Model, TextDetectionByEAST)
...
+    std::vector<float> keep_confidences;
+    std::vector<Rect2d> keep_boxes;
+    for (auto i : keep_idx)
+    {
+        keep_classIds.push_back(classIds[i]);
+        keep_confidences.push_back(confidences[i]);
+        keep_boxes.push_back(boxes[i]);
+    }
+
+    normAssertDetections(refClassIds, refScores, refBoxes, keep_classIds, keep_confidences, keep_boxes);
```
Comment (alalek) [inline modules/dnn/test/test_model.cpp:886]: There are 3 extra parameters to tune checks of this call.

Default values 1e-5/1e-4 is overkill for real networks.
Try to add this `..., "", 0, 0.01f, 0.01f`

**Comment by alalek** [comment]
It makes sense to add performance test too.
Here: https://github.com/opencv/opencv/blob/4.6.0/modules/dnn/perf/perf_net.cpp#L221

File: modules/dnn/test/test_model.cpp (hunk 7 above)
Comment (fengyuentau) [reply modules/dnn/test/test_model.cpp]: Now use png instead.

File: modules/dnn/test/test_model.cpp (hunk 8 above)
Comment (fengyuentau) [reply modules/dnn/test/test_model.cpp]: Test is now moved to `test_onnx_importer.cpp`.

File: modules/dnn/test/test_model.cpp (hunk 9 above)
Comment (fengyuentau) [reply modules/dnn/test/test_model.cpp]: If the issue persists with PNG, then I will change tolerance threshold.

**Comment by fengyuentau** [comment]
> It makes sense to add performance test too. Here: https://github.com/opencv/opencv/blob/4.6.0/modules/dnn/perf/perf_net.cpp#L221

It is indeed a good place for the test, but I think we keep it in `test_onnx_importer.onnx` since we need to check the result correctness.

**Comment by fengyuentau** [comment]
Hello @asenyaev , I observed that in Windows10-x64, the model that I manually uploaded to $dnn_models/dnn/onnx/models/yolov7-not-simplified.onnx got deleted somehow from time to time. But this issue did not happen on other hosts. Is it because of the scheduled update script or something like that?

File: modules/dnn/test/test_model.cpp (hunk 9 above)
Comment (fengyuentau) [reply modules/dnn/test/test_model.cpp]: Tests on default CI is now passed. Thank you for all the suggestions.

**Comment by alalek** [comment]
> since we need to check the result correctness.

Adding of performance test doesn't mean to drop the accuracy test.

**Comment by asenyaev** [comment]
Hello @fengyuentau! I checked logs in different workflows and want to say it's a strange behavior, the model was copied in the latest pipeline and DNN models weren't updated that day. Moreover, when models update, self-downloaded are not deleted. I see your model right there since September 6.

**Comment by fengyuentau** [comment]
Thank you @asenyaev . I guess I should just store the model somewhere else other than google drive, where our CI servers can download from, to avoid this kind of issue.

**Review by rogday — APPROVED** [review]
:+1:

**Comment by asmorkalov** [comment]
@fengyuentau Please sqush the commits.
